### PR TITLE
Task/short circuit literal node and literal subset node

### LINKED ISSUE #1524: Short Circuit Visitors - LiteralNode and LiteralSubsetNode
For any visitors that extend BaseVisitor, we have the option to short circuit at leaf nodes.

Improving the speed of our visitors will cut query planning time and, in some cases, speed up the time it takes to debug a query.

### PR DESCRIPTION
**Pull request by mineralntl**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java [hunk 1]
```diff
@@ -42,4 +45,47 @@ public Object visit(ASTEQNode node, Object data) {
...
     
+    // Descend through these nodes
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java]: The short circuit visitor does not short circuit and, or, reference, or expression nodes.  I don't think you need to override those methods here.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java [hunk 2]
```diff
@@ -35,4 +38,47 @@ public Object visit(ASTEQNode node, Object data) {
...
     
+    // Descend through these nodes
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java]: same comment as above.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java (hunk 1 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java]: For some reason, I thought BaseVisitor blocked but must have been an oversight.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java (hunk 2 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java]: For some reason, I thought BaseVisitor blocked but must have been an oversight.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java [hunk 3]
```diff
@@ -42,4 +45,23 @@ public Object visit(ASTEQNode node, Object data) {
...
     
+    // Descend through these nodes
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTERNode node, Object data) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java:53]: Do we need to implement this for the ASTNRNode too? (Is that the "not equals regex" node?)

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    
+    private final Set<String> expectedLiterals = new HashSet<>(Arrays.asList("FOO", "BAR", "BAZ", "ABCD", "1234", "1b3d"));
+    
+    @Test
+    public void testSingleLiteralWithMatch() throws ParseException {
+        String queryString = "FOO == '1234'";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(queryString);
+        Multimap<String,String> literals = LiteralNodeSubsetVisitor.getLiterals(expectedLiterals, script);
+        
+        Assert.assertEquals(1L, literals.size());
```
Comment (jschmidt10) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Nit: size() is returning an integer I think so this is probably auto-converting the size() to a long unnecessarily. We could just do `assertEquals(1, literals.size())`

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        Multimap<String,String> literals = LiteralNodeSubsetVisitor.getLiterals(expectedLiterals, script);
+        
+        Assert.assertEquals(2L, literals.size());
+        Assert.assertTrue(literals.containsEntry("FOO", "1234"));
+        Assert.assertTrue(literals.containsEntry("BAZ", "1b3d"));
+    }
+    
+    @Test
+    public void testMultipleLiteralsNoMatch() throws ParseException {
+        String queryString = "YO == 'HO' || YO == 'HO' && A == 'PIRATESLIFEOFRME'";
```
Comment (jschmidt10) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]:
```suggestion
        String queryString = "YO == 'HO' || YO == 'HO' && A == 'PIRATESLIFEFORME'";
```

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+public class LiteralNodeSubsetVisitorTest {
+    
+    private final Set<String> expectedLiterals = new HashSet<>(Arrays.asList("FOO", "BAR", "BAZ", "ABCD", "1234", "1b3d"));
+    
+    @Test
+    public void testSingleLiteralWithMatch() throws ParseException {
+        String queryString = "FOO == '1234'";
```
Comment (jschmidt10) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java:21]: We could simplify these unit tests with a helper method:

```
Multimap<String,String> literals = getLiterals("FOO == '1234'");
// make literal assertions
```

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 5 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Oversight - fixed.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 4 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Changed as suggested.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java (hunk 3 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitor.java:53]: Oversight on my part.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 6 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java:21]: Fair point - simplified as suggested.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java [hunk 7]
```diff
@@ -35,4 +38,23 @@ public Object visit(ASTEQNode node, Object data) {
...
     
+    // Descend through these nodes
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTERNode node, Object data) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java:46]: My bad, this one needs the ASTNRNode as well I think.

**Review by jschmidt10 — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java (hunk 7 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/LiteralNodeVisitor.java:46]: Updated

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        Multimap<String,String> literals = getLiterals(queryString);
+        
+        Assert.assertEquals(1, literals.size());
+        Assert.assertTrue(literals.containsEntry("FOO", "1234"));
+    }
+    
+    @Test
+    public void testSingleLiteralWithoutMatch() throws ParseException {
+        String queryString = "HERPDERP == '10000000'";
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: realistic field names please

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeVisitorTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
+package datawave.query.jexl.visitors;
+
+import com.google.common.collect.Multimap;
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ParseException;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class LiteralNodeVisitorTest {
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeVisitorTest.java:17]: Let's include some test cases that exercise functions (include, exclude at a minimum) and some marker nodes (value exceeded, term exceeded, delayed)

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ParseException;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class LiteralNodeSubsetVisitorTest {
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java:17]: This whole class could use a refactor. Delegate all tests to a single test method that looks like `test(String query, Map<String,String> expectedLiterals)`.

Then your test methods looks like
```
    @Test
    public void testSingleLiteralWithMatch() throws ParseException {
        String query = "FOO == '1234'";
        Multimap<String,String> literals = Multimaps.newArrayListMultimap();
        literals .put("FOO", "1234");
        test(query, literals );
    }
```

This comment applies to the other test.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    public void testMultipleLiteralsMixedMatchAndRegex() throws ParseException {
+        String queryString = "FOO == '1234' && FLIM == 'FLAM' || BAZ =~ 'abcd.*\\d'";
+        
+        Assert.assertEquals(1, getLiterals(queryString).size());
+        Assert.assertTrue(getLiterals(queryString).containsEntry("FOO", "1234"));
+    }
+    
+    @Test
+    public void testMultipleLiteralsNoMatchAndRegex() throws ParseException {
+        String queryString = "FLIM == 'FLAM' && PARA == 'diddle' || CHEESE =~ 'ka.*\\d'";
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Fields uppercased, values lowercased, please.

**Review by apmoriarty — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 8 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Fixed

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 11 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Fixed

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 10 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java:17]: Done and done.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    public void testIncludeRegex3() throws ParseException {
+        String query = "FOO == 'bar' || filter:includeRegex(FOO,'ba.*')";
+        Multimap<String,String> literals = ArrayListMultimap.create();
+        literals.put("FOO", "bar");
+        
+        test(query, literals);
+    }
+    
+    @Test
+    public void testFunctionInsideQueryPropertyMarker() throws ParseException {
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Rename this method to reflect the query, i.e., `testDelayedEqNode` or `testEqNodeInsideQueryPropertyMarker`

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java (hunk 12 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/LiteralNodeSubsetVisitorTest.java]: Done and done.

**Review by alerman — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]
